ДЗ 2, Кравченко Юрий, подгруппа 2

XJIE6 opened the pull request.

dancingrobot84 commented:
Почему тесты на Groovy? Мы же вроде Java изучаем.

dancingrobot84 commented:
Также сообщи на email или оставь здесь комментарий когда можно будет проверять работу.

XJIE6 commented:
@dancingrobot84 Подразумевалось, что уже можно проверять. Тесты поменял. IntelliJ по дефолту Groovy делает. Думал, так и надо...

dancingrobot84 commented on src/main/java/ru/spbau/mit/Collections.java:
```diff
@@ -0,0 +1,80 @@
+ */
+public class Collections {
+
+    public static <T, R> ArrayList<R> map(Function1<? super T, ? extends R> func, ArrayList<T> collection) {
```
В задании написано использовать интерфейс `Iterable` для коллекций. Однако, даже если бы этого не было написано стоило использовать интерфейс `List` вместо конкретной реализации `ArrayList`.

dancingrobot84 commented on src/main/java/ru/spbau/mit/Collections.java:
```diff
@@ -0,0 +1,80 @@
+        if (!iter.hasNext()) {
+            return start;
+        }
+        return foldr(func, func.apply(start, iter.next()), iter);
```
Упадет с переполнением стека для коллекций размером больших чем стек. Переделай без рекурсии.

dancingrobot84 commented on src/main/java/ru/spbau/mit/Collections.java:
```diff
@@ -0,0 +1,80 @@
+            return start;
+        }
+        T2 val = iter.next();
+        return func.apply(foldl(func, start, iter), val);
```
Упадет с переполнением стека для коллекций размером больших чем стек. Переделай без рекурсии.

dancingrobot84 commented on src/main/java/ru/spbau/mit/Function1.java:
```diff
@@ -0,0 +1,18 @@
+abstract public class Function1<T, R> {
+    public abstract R apply(T arg);
+
+    public <R2> Function1<T, R2> compose(final Function1<? super R, R2> outer) {
```
Не самая гибкая сигнатура для `outer`.

dancingrobot84 commented on src/main/java/ru/spbau/mit/Function2.java:
```diff
@@ -0,0 +1,47 @@
+abstract public class Function2<T, T2, R>{
+    public abstract R apply(T arg, T2 arg2);
+
+    public <R2> Function2<T, T2, R2> compose(final Function1<? super R, R2> outer) {
```
Не самая гибкая сигнатура для `outer`.

dancingrobot84 commented on src/test/java/ru/spbau/mit/CollectionsTest.java:
```diff
@@ -0,0 +1,104 @@
+        while (it.hasNext() && it2.hasNext()) {
+            assertEquals(it.next(), it2.next());
+        }
+        assertEquals(it.hasNext(), it2.hasNext());
```
Код для проверки одинаковости итераторов нужно вынести в отдельную функцию

dancingrobot84 commented on src/test/java/ru/spbau/mit/CollectionsTest.java:
```diff
@@ -0,0 +1,104 @@
+    }
+    @Test
+    public void testFoldl() throws Exception {
+        assertEquals("223454321", Collections.foldl(concat, "", arr));
```
Поскольку конкатенация строк ассоциативна, то этот тест бесполезен, т.к. неважно что использовать, `foldl` или `foldr`. Но, поскольку у тебя он еще и возвращает разные результаты, значит `foldl` написан неверно. Надо:
1. Поправить `foldl`
2. Придумать более показательный тест

dancingrobot84 commented on src/test/java/ru/spbau/mit/PredicateTest.java:
```diff
@@ -0,0 +1,48 @@
+        assertFalse(eqTen.apply(0));
+    }
+    @Test
+    public void testOr() throws Exception {
```
Не проверяется ленивость `or`

dancingrobot84 commented on src/test/java/ru/spbau/mit/PredicateTest.java:
```diff
@@ -0,0 +1,48 @@
+        assertFalse(eqTen.or(eqTen).apply(9));
+    }
+    @Test
+    public void testAnd() throws Exception {
```
Не проверяется ленивость `and`

dancingrobot84 commented:
@XJIE6 проверил

XJIE6 commented:
@dancingrobot84 
1)fixed
2)fixed
3)fixed
4)fixed
5)fixed
6)fixed
7)Мне кажется, что всё правильно. В одном случае concat с начала, в другом случае с конца и получается перевёрнутая строка
8)fixed
9)fixed

dancingrobot84 commented:
@XJIE6 

Нет, foldr и foldl не про начало и конец, а про расстановку скобок. В твоем примере выходные строки должны совпадать потому, что конкатенация ассоциативна, а значит нет разницы между:

```
foldl(+, "", ["1", "2", "3"]) == ((("" +"1") + "2") + "3") == "123"
foldr(+, "", ["1", "2", "3"]) == ("" + ("1" + ("2" + "3"))) == "123"
```

Почитай подробнее описание свертки [на википедии](https://ru.wikipedia.org/wiki/%D0%A1%D0%B2%D1%91%D1%80%D1%82%D0%BA%D0%B0_%D1%81%D0%BF%D0%B8%D1%81%D0%BA%D0%B0)

dancingrobot84 commented:
Не принято: не исправлен foldl/foldr.
